components/tooltip/script.js

```diff
@@ -65,91 +78,114 @@ class Tooltip extends HTMLElement {
+     * and checks if the tooltip is visible on the new position. If it is not - calls setPosition
+     * recursively until a suitable position is found or we've ran out of possibilities.
+     * @param {number} scrollOffsetX - the offset of the scroll on the X axis.
+     * @param {number} scrollOffsetY - the offset of the scroll on the Y axis.
+     * @param {string} orientation - the current position of the tooltip - top, left, right or bottom.
+     * @returns {promise}
+    */
+    async setPosition(scrollOffsetX, scrollOffsetY, orientation = this.position) {
+        const elementSize = this.triggerElement.getBoundingClientRect();
 
-        return overflows;
-    }
+        let [updatedOrientation, elementPosition] = this.getPositionCoords(orientation, elementSize);
 
-    getVisiblePosition(overflows) {
-        const allSides = ['top', 'bottom', 'left', 'right'];
+        this.position = updatedOrientation;
+        this.style.top = scrollOffsetY + elementPosition.top + 'px';
+        this.style.left = scrollOffsetX + elementPosition.left + 'px';
 
-        overflows = overflows || this.overflows();
+        await this.waitForFrames(2);
+        const overflows = this.overflows;
         let overflowingSides = Object.keys(overflows);
+        if (overflowingSides.length && this.uncheckedOrientations.length !== 0) {
+            return await this.setPosition(scrollOffsetX, scrollOffsetY, this.getVisiblePosition(overflowingSides));
+        }
 
-        overflowingSides = overflowingSides.filter(side => this.fixedSides.indexOf(side) == -1);
+        return Promise.resolve();
+    }
 
-        let nextIndex = allSides.indexOf(this.position) + 1;
-        nextIndex = [(nextIndex + allSides.length) % allSides.length];
-        this.position = allSides[nextIndex];
-        this.fixedSides.push(this.position);
+    /**
+     * An asynchronous method that waits for given amount of frames before it resolves;
+     * useful in cases where we need to wait for styles to be computed.
+     * @param {number} count - the number of frames it should wait for before resolving.
+     * @returns {promise}
+     */
+    async waitForFrames(count) {
+        while(count--) {
+            await new Promise((resolve) => requestAnimationFrame(resolve));
+        }
+        return Promise.resolve();
+    }
 
-        return this.position;
+    /**
+     * Gets the next value of a list of possible positions for the tooltip.
+     * The possible positions is the difference between the overflowing positions and the
+     * positions that haven't been tried yet.
+     * @param {Array<string>} overflowingSides - an array of the sides of which the tooltip is hidden.
+     * @returns {string} position - the new position.
+     */
+    getVisiblePosition(overflowingSides) {
+        this.uncheckedOrientations = this.uncheckedOrientations.filter(side => !overflowingSides.includes(side));
+        let position = this.uncheckedOrientations[0];

```

Comment:
```suggestion
        const position = this.uncheckedOrientations.shift();

```
And you can remove the next line

Reply:
Nice - I didn't know shift returned the removed element. I also removed the variable, we now directly return `this.uncheckedOrientations.shift();`. Also added a comment just in case to avoid confusion.